File: Reqnroll/Chapter15/15.2-TestDriverPattern/Before/WIMP.Specs/StepDefinitions/AuthenticationStepDefinitions.cs
using System.Net;
using System.Net.Http.Json;

using Reqnroll;

using WIMP.App.RestApi;
using WIMP.Specs.Support;

namespace WIMP.Specs.StepDefinitions;

[Binding]
public class AuthenticationStepDefinitions(AuthenticationContext authContext, RestApiContext restApiContext, AppHostingContext appHostingContext)
{
    private HttpResponseMessage? loginApiResponse;

    [Given("the customer has authenticated")]
    public async Task GivenTheCustomerHasAuthenticated()
    {
        var payload = new LoginRequest(
            DomainDefaults.CustomerName,
            DomainDefaults.Password);
        var response = await appHostingContext.AppHost.CreateClient()
            .PostAsJsonAsync("/api/auth/login", payload);

        Assert.AreEqual(HttpStatusCode.OK, response.StatusCode);
        var loginResponse = await response.Content.ReadFromJsonAsync<LoginResponse>()
                            ?? throw new InvalidOperationException("No result payload found");
        restApiContext.BearerToken = loginResponse.Token;
        authContext.LoggedInCustomerName = DomainDefaults.CustomerName;
    }

    [When("the customer attempts to log in with a wrong password")]
    public async Task WhenTheCustomerAttemptsToLogInWithAWrongPassword()
    {
        var payload = new LoginRequest(
            DomainDefaults.CustomerName,
            DomainDefaults.WrongPassword);
        loginApiResponse = await appHostingContext.AppHost.CreateClient()
            .PostAsJsonAsync("/api/auth/login", payload);
    }

    [Then("the login should fail with {string}")]
    public async Task ThenTheLoginShouldFailWith(string expectedMessage)
    {
        Assert.IsNotNull(loginApiResponse);
        Assert.AreNotEqual(HttpStatusCode.OK, loginApiResponse.StatusCode, "Login should fail");
        string errorMessage = await loginApiResponse.Content.ReadAsStringAsync();
        StringAssert.Contains(errorMessage, expectedMessage,
            "Login should fail with the right error message");
    }
}

File: CucumberJVM/Chapter15/15.6-PolymorphicDriverPattern/After/src/test/java/com/wimp/app/specs/stepdefinitions/AuthenticationStepDefinitions.java
package com.wimp.app.specs.stepdefinitions;

import com.wimp.app.restapi.LoginResponse;
import com.wimp.app.specs.drivers.AuthenticationDriver;
import com.wimp.app.specs.support.DomainDefaults;
import com.wimp.app.specs.support.TestActionResult;
import io.cucumber.java.en.*;

public class AuthenticationStepDefinitions {
    private final AuthenticationDriver authenticationDriver;
    private TestActionResult<LoginResponse> loginResult;

    public AuthenticationStepDefinitions(AuthenticationDriver authenticationDriver) {
        this.authenticationDriver = authenticationDriver;
    }

    @When("the customer attempts to log in with a wrong password")
    public void theCustomerAttemptsToLogInWithAWrongPassword() throws Exception {
        loginResult = authenticationDriver.login(DomainDefaults.CUSTOMER_NAME, DomainDefaults.WRONG_PASSWORD).attemptExecute();
    }

    @Then("the login should fail with {string}")
    public void theLoginShouldFailWith(String expectedMessage) {
        loginResult.assertFailedWithErrorMessageContains(expectedMessage);
    }
}
